### PR TITLE
fix and optimize erroneous code paths

### PR DESCRIPTION
**Pull request by aviatesk**

### CODE REVIEW

File: src/syntax_tree.jl [hunk 1]
```diff
@@ -118,13 +118,12 @@ function interpolate_literal(node::SyntaxNode, val)
...
-function _show_syntax_node(io, current_filename, node, indent)
+function _show_syntax_node(io, current_filename, node::SyntaxNode, indent)
     fname = node.source.filename
     line, col = source_location(node.source, node.position)
     posstr = "$(lpad(line, 4)):$(rpad(col,3))│$(lpad(node.position,6)):$(rpad(node.position+span(node)-1,6))│"
-    nodestr = haschildren(node)   ?  "[$(untokenize(head(node)))]" :
-              node.val isa Symbol ? string(node.val) :
-              repr(node.val)
+    nodestr = haschildren(node) ? "[$(untokenize(head(node)))]" :
+              (val = node.val; isa(val, Symbol)) ? string(val) : repr(val)
```
Comment (pfitzseb) [inline src/syntax_tree.jl:126]:
```suggestion
    val = node.val
    nodestr = haschildren(node) ? "[$(untokenize(head(node)))]" :
              isa(val, Symbol) ? string(val) : repr(val)
```

**Review by pfitzseb — APPROVED** [review]

File: src/syntax_tree.jl [hunk 2]
```diff
@@ -141,12 +141,13 @@ function _show_syntax_node(io, current_filename, node, indent)
...
 
-function _show_syntax_node_sexpr(io, node)
+function _show_syntax_node_sexpr(io, node::SyntaxNode)
     if !haschildren(node)
         if is_error(node)
             print(io, "(", untokenize(head(node)), ")")
         else
-            print(io, node.val isa Symbol ? string(node.val) : repr(node.val))
+            val = node.val
+            print(io, val isa Symbol ? string(val) : repr(val))
```
Comment (c42f) [inline src/syntax_tree.jl:150]: I'm curious - is this stylistic or does it do something more?

Ah wait I think I see - this allows inference to prove that `val::Symbol` inside the first branch.

File: src/value_parsing.jl [hunk 3]
```diff
@@ -228,7 +228,7 @@ function utf8proc_decompose(str, options, buffer, nwords)
     ret = ccall(:utf8proc_decompose_custom, Int, (Ptr{UInt8}, Int, Ptr{UInt8}, Int, Cint, Ptr{Cvoid}, Ptr{Cvoid}),
                 str, sizeof(str), buffer, nwords, options,
                 @cfunction(utf8proc_custom_func, UInt32, (UInt32, Ptr{Cvoid})), C_NULL)
-    ret < 0 && utf8proc_error(ret)
+    ret < 0 && Base.Unicode.utf8proc_error(ret)
```
Comment (c42f) [inline src/value_parsing.jl:231]: oops. Thanks!

File: src/diagnostics.jl [hunk 4]
```diff
@@ -32,7 +32,7 @@ end
 function Diagnostic(first_byte, last_byte; error=nothing, warning=nothing)
     message = !isnothing(error)   ? error :
               !isnothing(warning) ? warning :
-              error("No message in diagnostic")
+              Base.error("No message in diagnostic")
```
Comment (c42f) [inline src/diagnostics.jl:35]: Oops!  Thanks for the fix.

**Review by c42f — APPROVED** [review]
This looks great, thanks. Feel free to merge it!

File: src/syntax_tree.jl (hunk 2 above)
Comment (aviatesk) [reply src/syntax_tree.jl:150]: Yeah currently inference can propagate conditional type information about local variable but not about memory owned by local variable. So we need to extract the field value into a local variable to get that extra type propagation.

**Comment by aviatesk** [comment]
> Feel free to merge it!

Thanks! I don't have a write access to this repository so please merge this👍

**Comment by c42f** [comment]
Oh, I thought you'd have access as a member of the JuliaLang/committers group.  I'd add you to this repo directly but I seem to have lost admin when I transferred this repo (requesting access again now...)

For now I'll just merge this. Thanks @pfitzseb for the review :)

**Pull request merged**